File: src/scale.c
#include <libavutil/avutil.h>
#include <libavutil/imgutils.h>
#include <libswscale/swscale.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// https://github.com/obsproject/obs-studio/blob/542cb876dce6077bb24fa5ee0439e7d21ac53f59/libobs/media-io/video-scaler-ffmpeg.c#L148

struct Scaler {
  struct SwsContext *swsContext;
};

#define FIXED_1_0 (1 << 16)
void *new_scaler(int srcW, int srcH, enum AVPixelFormat srcFormat,
                 bool srcFullRange, int srcSpace, int dstW, int dstH,
                 enum AVPixelFormat dstFormat, bool dstFullRange,
                 int dstSpace) {
  struct Scaler *scaler = NULL;
  struct SwsContext *swsContext = NULL;

  swsContext = sws_alloc_context();
  if (!swsContext) {
    fprintf(stderr, "Failed to alloc SwsContext\n");
    goto _exit;
  }
  av_opt_set_int(swsContext, "sws_flags", SWS_FAST_BILINEAR, 0);
  av_opt_set_int(swsContext, "srcw", srcW, 0);
  av_opt_set_int(swsContext, "srch", srcH, 0);
  av_opt_set_int(swsContext, "dstw", dstW, 0);
  av_opt_set_int(swsContext, "dsth", dstH, 0);
  av_opt_set_int(swsContext, "src_format", srcFormat, 0);
  av_opt_set_int(swsContext, "dst_format", dstFormat, 0);
  av_opt_set_int(swsContext, "src_range", srcFullRange ? 1 : 0, 0);
  av_opt_set_int(swsContext, "dst_range", dstFullRange ? 1 : 0, 0);

  if (sws_init_context(swsContext, NULL, NULL) < 0) {
    fprintf(stderr, "Failed to init SwsContext\n");
    goto _exit;
  }

  if (sws_setColorspaceDetails(
          swsContext, sws_getCoefficients(srcSpace), srcFullRange ? 1 : 0,
          sws_getCoefficients(srcSpace), dstFullRange ? 1 : 0, 0, FIXED_1_0,
          FIXED_1_0) < 0) {
    fprintf(stderr, "Failed to set colorspace details\n");
    goto _exit;
  }

  scaler = (struct Scaler *)malloc(sizeof(struct Scaler));
  if (!scaler) {
    fprintf(stderr, "Failed to alloc scaler\n");
    goto _exit;
  }
  scaler->swsContext = swsContext;
  return (void *)scaler;

_exit:
  if (swsContext) {
    sws_freeContext(swsContext);
  }
}

int scale(const void *const scaler, const uint8_t *const srcSlice[],
          const int srcStride[], int srcSliceY, int srcSliceH,
          uint8_t *const dst[], const int dstStride[]) {
  struct SwsContext *swsContext = ((struct Scaler *)scaler)->swsContext;
  return sws_scale(swsContext, srcSlice, srcStride, srcSliceY, srcSliceH, dst,
                   dstStride);
}

void free_scaler(void *scaler) {
  if (!scaler) {
    return;
  }
  struct SwsContext *swsContext = ((struct Scaler *)scaler)->swsContext;
  if (swsContext) {
    sws_freeContext(swsContext);
  }
}

File: src/scale.rs
use log::{error, trace};

use crate::{av_log_get_level, free_scaler, new_scaler, scale, AVPixelFormat, AV_LOG_ERROR};
use std::{ffi::c_void, os::raw::c_int};

#[derive(Debug, Clone, PartialEq)]
pub struct ScaleContext {
    pub srcW: c_int,
    pub srcH: c_int,
    pub srcFormat: AVPixelFormat,
    pub srcFullRange: bool,
    pub srcSpace: c_int,
    pub dstW: c_int,
    pub dstH: c_int,
    pub dstFormat: AVPixelFormat,
    pub dstFullRange: bool,
    pub dstSpace: c_int,
}

pub struct Scaler {
    inner: Box<c_void>,
    pub ctx: ScaleContext,
}

unsafe impl Send for Scaler {}
unsafe impl Sync for Scaler {}

impl Scaler {
    pub fn new(ctx: ScaleContext) -> Result<Self, ()> {
        unsafe {
            let inner = new_scaler(
                ctx.srcW,
                ctx.srcH,
                ctx.srcFormat,
                ctx.srcFullRange,
                ctx.srcSpace,
                ctx.dstW,
                ctx.dstH,
                ctx.dstFormat,
                ctx.dstFullRange,
                ctx.dstSpace,
            );
            if inner.is_null() {
                return Err(());
            }

            Ok(Scaler {
                inner: Box::from_raw(inner as *mut c_void),
                ctx,
            })
        }
    }

    pub fn scale(
        &self,
        srcSlice: *const *const u8,
        srcStride: *const c_int,
        srcSliceY: c_int,
        srcSliceH: c_int,
        dst: *const *mut u8,
        dstStride: *const c_int,
    ) -> Result<(), i32> {
        unsafe {
            let result = scale(
                &*self.inner,
                srcSlice,
                srcStride,
                srcSliceY,
                srcSliceH,
                dst,
                dstStride,
            );
            if result < 0 {
                if av_log_get_level() >= AV_LOG_ERROR as _ {
                    error!("Error scale: {}", result);
                }
                return Err(result);
            }
            Ok(())
        }
    }
}

impl Drop for Scaler {
    fn drop(&mut self) {
        unsafe {
            free_scaler(self.inner.as_mut());
            trace!("Scaler dropped");
        }
    }
}
